### PR TITLE
[SDEM] Fluid sdem coupling implementation

### PR DESCRIPTION
**Pull request by jgonzalezusua**
**Description**
The aim of this pull request is to couple the _SwimmingDEMApplication_ and _FluidDynamicsApplication_ to be able to solve fluid-particle _"two-way"_  coupling problems

Please mark the PR with appropriate tags: 
- Feature, Application

**Changelog**
Please summarize the changes in one list to generate the changelog:
- Added a derived solver from _navier_stokes_vmsmonolithic_solver.py_.
- Added processes to compute manufactured solutions in cpp and the python modules.
- Added process to compute exact_variables.
- Added an element to calculate the L2 projection error.
- Added a _"two-way"_ SDEM-Fluid test to compute the fluid fraction.
- Added a L2 projection utility.
- Added a python script that writes in hdf5 format the error.
- Added derived schemes from bossak and bdf2 located in the _FluidDynamicsApplication_ to calculate the fluid_fraction_rate.
- Added a python analysis and solver for manufactured cases.

### CODE REVIEW

**Comment by philbucher** [comment]
@jgonzalezusua can you please reduce the size of the mdpa files that you are adding
If possible we try to keep it below 1000 lines

also please squash this PR instead of merging to avoid having the large files in the git history

Thx

**Comment by jgonzalezusua** [comment]
I've merged the master branch to this one to fix the compilation errors related with the atomic_utilities

**Comment by jgonzalezusua** [comment]
@philbucher what do you mean about squash the PR?

**Comment by maceligueta** [comment]
It’s an optional type of merge. You can collapse all commits in one. Just select from the drop down merge menu (I never used it so far).

**Comment by philbucher** [comment]
I will explain in more detail later, also I have some comments, please wait with merging

File: applications/SwimmingDEMApplication/custom_processes/spatial_dependant_porosity_solution_body_force_process.cpp [hunk 1]
```diff
@@ -12,12 +12,12 @@
 //
 
 // System includes
-#include <math.h>
+#define _USE_MATH_DEFINES
+#include <cmath.h>
```
Comment (philbucher) [inline applications/SwimmingDEMApplication/custom_processes/spatial_dependant_porosity_solution_body_force_process.cpp]: please use [this](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/includes/global_variables.h#L36) instead

**Comment by philbucher** [comment]
> @philbucher what do you mean about squash the PR?

I meant [this](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/merging-a-pull-request#merging-a-pull-request-on-github)

The idea is that you compress/squash all the commits of a PR into a single commit, aka only the actual changes to the code will be merged to master
the reason I am asking you to do this is to reduce the size of the git history of Kratos. Since you committed large mdpa files they are in the history now. Even if you delete them again, they will remain there forever.
Hence I am asking you to reduce their size and use a squash merge, such that the large files don't end up in the history

To give some perspective as to why this is important: the Kratos source code is ~300MB, the git history is over 2GB! => in the past large files were committed (and removed in the meantime), but they remain in the history forever and it is basically impossible to remove them

File: applications/SwimmingDEMApplication/python_scripts/hdf5_script.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
+# Import system python
+import os
+
+import numpy as np
+import h5py
+import KratosMultiphysics as Kratos
+from KratosMultiphysics import Vector
+import KratosMultiphysics.SwimmingDEMApplication as Dem
+
+class ErrorProjectionPostProcessTool(object):
```
Comment (philbucher) [inline applications/SwimmingDEMApplication/python_scripts/hdf5_script.py:9]: FYI I the hdf5application would probably work for you too

**Comment by jgonzalezusua** [comment]
Seems like Codacy has problems with the docstrings and blocks the PR. I am a little confused because other PR have similar docstrings format but Codacy aproves them. @rogicarlo, could you help me with that?

If you look, for example, in the Codacy review of this PR #7620 . You will notice that in _poromechanics_U_Pw_solver.py_ Codacy shows the same errors as in some commits I pushed before. But, at the end of the day, it approves that PR and blocked mine.

I do not understand this behavior.

**Comment by roigcarlo** [comment]
Hi @jgonzalezusua , codacy does not block the PR, but all CI are failing. Seems to be an error while loading the mmpg library (`libmmg2d.so.5`):

```shell
ImportError: libmmg2d.so.5: cannot open shared object file: No such file or directory
```
@philbucher Does this has something to do with the tests you are doing on #7622, etc? Seems that pretty much all PR after that one are failing :S

**Comment by philbucher** [comment]
yes this is fixed in #7622 
sorry for the inconvenience

**Comment by jgonzalezusua** [comment]
@roigcarlo sorry for the misunderstanding when I said that Codacy is blocking. I mean it says that my code is not up to standards even if in other PRs have the same docstring format and it approves them.

Thanks @philbucher for solving the issue with the tests

**Comment by roigcarlo** [comment]
A ok, I misunderstud. The problem is that its only going to report files changed in this commit, not previously existing ones, #7620 only touches `cpp` and `h` files, so only reports there.

As for the "errors", is just styling:

```python
def function():
    """Multiline
    multiline 2
    """
```

to

```python
def function():
    """
    Multiline

    multiline 2
    """
```

But technically speaking both are ok, one is called `D212` and the other `D213`, and are mutually exclusive. It just happens that the one you are using and most of the files in Kratos had is `D212`. 

We will try to change it so it does not give those problems.

File: applications/SwimmingDEMApplication/custom_elements/calculate_error_L2_projection_element.h [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+
+///@}
+///@name  Functions
+///@{
+
+///@}
+///@name Kratos Classes
+///@{
+
+/// A post-processing element to recover the Laplacian from the velocity solution.
```
Comment (GuillermoCasas) [inline applications/SwimmingDEMApplication/custom_elements/calculate_error_L2_projection_element.h]: Change this comment

File: applications/SwimmingDEMApplication/custom_strategies/strategies/bdf2_turbulent_schemeDEMCoupled.h [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+        double delta_time = r_current_process_info[DELTA_TIME];
+        double delta_time_inv = 1.0 / delta_time;
+
+        //for(int k = 0; k<static_cast<int>(rModelPart.Nodes().size()); k++)
+        block_for_each(r_model_part.Nodes(), [&](Node<3>& rNode)
+        {
+            const double fluid_fraction = rNode.FastGetSolutionStepValue(FLUID_FRACTION);
+            const double fluid_fraction_old = rNode.FastGetSolutionStepValue(FLUID_FRACTION_OLD);
+            const double fluid_fraction_rate = delta_time_inv * (fluid_fraction - fluid_fraction_old);
+            rNode.FastGetSolutionStepValue(FLUID_FRACTION_RATE) = fluid_fraction_rate;
```
Comment (GuillermoCasas) [inline applications/SwimmingDEMApplication/custom_strategies/strategies/bdf2_turbulent_schemeDEMCoupled.h]: I think this should be made coherent with the bdf2 scheme... It should be easy by copying what is done in InitializeSolutionStep from the original bdf2 scheme

File: applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+                error_y += rGeom[j].FastGetSolutionStepValue(ERROR_Y) * N[j];
+                if (dim == 3) error_z += rGeom[j].FastGetSolutionStepValue(ERROR_Z) * N[j];
+            }
+            error.push_back(error_x);
+            error.push_back(error_y);
+            if (dim == 3) error.push_back(error_z);
+            // for (unsigned int d = 0; d < dim; ++d){
+            //     scalar_product[d] = error[d];
+            // }
+            result += pow(SWIMMING_MODULUS_3(error),2) * DetJ[gss] * IntegrationPoints[gss].Weight();
```
Comment (GuillermoCasas) [inline applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h]: use std::pow.

File: applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+        for (SizeType gss = 0; gss < NumGauss; ++gss){
+            array_1d<double, 4> N;
+            N = row(NContainer, gss);
+            Vector DetJ = ZeroVector(NumGauss);
+            ielem->GetGeometry().DeterminantOfJacobian(DetJ, GeometryData::GI_GAUSS_2);
+            for (unsigned int j = 0; j < NumNodes; ++j){
+                error += rGeom[j].FastGetSolutionStepValue(SCALAR_ERROR) * N[j];
+            }
+            result += pow(error,2) * DetJ[gss] * IntegrationPoints[gss].Weight();
```
Comment (GuillermoCasas) [inline applications/SwimmingDEMApplication/custom_utilities/L2_error_projection_utility.h]: use std:pow

File: applications/SwimmingDEMApplication/python_scripts/hdf5_script.py (hunk 2 above)
Comment (GuillermoCasas) [reply applications/SwimmingDEMApplication/python_scripts/hdf5_script.py:9]: > FYI I the hdf5application would probably work for you too

Yes, we want to migrate all our h5py python tools to it if possible. We will discuss it for a future PR

File: applications/SwimmingDEMApplication/python_scripts/variables_management.py [hunk 7]
```diff
@@ -339,7 +339,7 @@ def ConstructListsOfVariablesForCoupling(self, parameters):
 
         # fluid coupling variables
         self.coupling_fluid_vars = []
-        self.coupling_fluid_vars += [Kratos.MATERIAL_ACCELERATION]
+        self.coupling_fluid_vars += [Kratos.MATERIAL_ACCELERATION, SDEM.EXACT_VELOCITY, SDEM.VECTORIAL_ERROR, SDEM.ERROR_X, SDEM.ERROR_Y, SDEM.ERROR_Z, SDEM.SCALAR_ERROR, SDEM.EXACT_PRESSURE, Kratos.MASS_SOURCE]
```
Comment (GuillermoCasas) [inline applications/SwimmingDEMApplication/python_scripts/variables_management.py]: use several lines

**Review by GuillermoCasas — CHANGES_REQUESTED** [review]
Please address comments

File: applications/SwimmingDEMApplication/custom_strategies/strategies/bdf2_turbulent_schemeDEMCoupled.h (hunk 4 above)
Comment (GuillermoCasas) [reply applications/SwimmingDEMApplication/custom_strategies/strategies/bdf2_turbulent_schemeDEMCoupled.h]: This is still not bdf2, right?

Comment (jgonzalezusua) [reply applications/SwimmingDEMApplication/custom_strategies/strategies/bdf2_turbulent_schemeDEMCoupled.h]: You have to press F5 to see the changes and go to the files changed in the commit before the last one to see the modifications in the bdf2

**Review by GuillermoCasas — APPROVED** [review]

**Comment by loumalouomega** [comment]
@jgonzalezusua can we delete the branch?

**Pull request merged**